File: design-principles/single-responsibility/go/SingleResponsibilityBadExample.go
/**
 * 单一职责原则 - 反例
 * 这个例子违反了单一职责原则。
 * 1. 订单处理类实现了订单校验以及保存数据库的两种逻辑。
 * 2. 一旦订单条件有修改或保存数据库方式有变更都需要改动此类。
 * 对比：正例将校验与数据持久化拆分为独立类，职责清晰、修改互不影响。
 */
package main

import "fmt"

type OrderProcessor struct{}

// 违规点：业务处理、校验、持久化混在一个结构体
func (p OrderProcessor) processOrder(orderId int) bool {
  fmt.Printf("oder ID：%d\n", orderId)

  // 校验逻辑应拆分为独立模块
  if !p.validateId(orderId) {
    fmt.Println("order validate id failed.")
    return false
  }

  if !p.validateTime(0) {
    fmt.Println("order validate time failed.")
    return false
  }

  if orderId%2 == 0 {
    fmt.Println("order data processing.")
  }

  // 持久化职责不应放在业务处理逻辑里
  fmt.Println("order save to DB.")
  p.saveOrder(orderId)
  return true
}

// 校验职责应独立
func (p OrderProcessor) validateId(orderId int) bool {
  return orderId%2 == 0
}

func (p OrderProcessor) validateTime(_ int) bool {
  return true
}

// 保存数据属于持久化职责
func (p OrderProcessor) saveOrder(orderId int) bool {
  if orderId%2 == 0 {
    fmt.Println("order saving.")
  }
  fmt.Println("order save done.")
  return true
}

func (p OrderProcessor) deleteOrder(_ int) bool {
  return true
}

func main() {
  processor := OrderProcessor{}
  processor.processOrder(1001)
  processor.processOrder(1002)
}

/**
jarry@jarrys-MBP go % go run SingleResponsibilityBadExample.go 
oder ID：1001
order validate id failed.
oder ID：1002
order data processing.
order save to DB.
order saving.
order save done.
*/

File: design-principles/single-responsibility/c/single_responsibility_bad_example.c
#include <stdio.h>
/**
 * 单一职责原则 - 反例
 * 这个例子违反了单一职责原则。
 * 1. 订单处理类实现了订单校验以及保存数据库的两种逻辑。
 * 2. 一旦订单条件有修改或保存数据库方式有变更都需要改动此类。
 * 对比：正例将校验与数据持久化拆分为独立类，职责清晰、修改互不影响。
 */

typedef struct {
  int unused;
} OrderProcessor;

int validate_id(int order_id) {
  return order_id % 2 == 0;
}

int validate_time(long _time) {
  return 1;
}

// 保存数据属于持久化职责
int save_order(int order_id) {
  if (order_id % 2 == 0) {
    printf("order saving.\n");
  }
  printf("order save done.\n");
  return 1;
}

// 违规点：一个模块负责校验、业务处理、持久化
int process_order(OrderProcessor *processor, int order_id) {
  (void)processor;
  printf("oder ID：%d\n", order_id);

  // 校验职责应独立
  if (!validate_id(order_id)) {
    printf("order validate id failed.\n");
    return 0;
  }

  if (!validate_time(0)) {
    printf("order validate time failed.\n");
    return 0;
  }

  if (order_id % 2 == 0) {
    printf("order data processing.\n");
  }

  // 持久化职责不应放在这里
  printf("order save to DB.\n");
  save_order(order_id);
  return 1;
}

int main() {
  OrderProcessor processor = {0};
  process_order(&processor, 1001);
  process_order(&processor, 1002);
  return 0;
}
/**
 * 单一职责原则 - 反例
 * 这个例子违反了单一职责原则。
 * 1. 订单处理类实现了订单校验以及保存数据库的两种逻辑。
 * 2. 一旦订单条件有修改或保存数据库方式有变更都需要改动此类。
 * 对比：正例将校验与数据持久化拆分为独立类，职责清晰、修改互不影响。
 */
